CL2-6833 i1 BE Get admin_publications publication_status counts

jinjagit opened the pull request:
## Checklist

- [ ] Added entry to changelog

NO CHANGELOG ENTRY - this PR does not fully implement any new feature.

- [x] Tests
<details>
<summary>More info</summary>

### Unit tests

Did you add relevant unit tests?

### E2E tests

Sometimes it can be more efficient to update E2E tests after CI has run them. If you know which ones to update, go ahead! E2E template cl2-back:

```bash
docker compose run --rm web bin/rails cl2_back:create_tenant[localhost,e2etests_template]
```

</details>

- [x] Prepared branch for code review
<details>
<summary>More info</summary>
Reviewed code to reduce unnecessary back and forth (removal of console.log, comments, ...)? Added comments to clarify code, emphasize what to pay attention to, etc.?
</details>

## Links

- [Jira ticket](https://citizenlab.atlassian.net/browse/CL2-6833)
- [Jira Epic](https://citizenlab.atlassian.net/browse/CL2-6832)

## What changes are in this PR?

A new endpoint, `status_counts`, in `AdminPublicationsController`, for use by the Homepage to get the counts of visible top-level `admin_publications` (representing folders and projects) by their `publication_status`, as part of the [Homepage Filters Epic](https://citizenlab.atlassian.net/browse/CL2-6832).  
  
The counts will be displayed on the homepage and also used in FE logic to decide whether to show related tabs.
   
<details>
  <summary>Example requests & responses - Click to expand</summary>

With these projects and folders on a platform: 
<img width="375" alt="Screenshot 2021-10-28 at 15 30 40" src="https://user-images.githubusercontent.com/3944042/139277814-7c671ed2-0d73-4e95-a190-6830b37ca71c.png">
As Admin:
<img width="1378" alt="Screenshot 2021-10-28 at 15 29 36" src="https://user-images.githubusercontent.com/3944042/139457956-d7406f49-cf8c-4f38-8dc3-c89afff0b981.png">
As non-admin User:
<img width="1378" alt="Screenshot 2021-10-28 at 15 30 03" src="https://user-images.githubusercontent.com/3944042/139457985-306de829-d8e2-4787-921a-1135f38681a5.png">
</details>
  
Why a new endpoint (and not just derive the counts from the publications data returned to FE)?
- [Faster db query when only ask for counts]( https://stackoverflow.com/questions/15912749/is-count-faster-than-pulling-the-records-and-counting-in-code)
- Smaller response size (obviously)
- More flexible: Can be used in parallel with other requests, for example enabling counts to be displayed before all related publications data is received, processed and displayed by FE. Can be used standalone if any reason arises to only get counts without related publications data.

### Filtering Service Context
[Study of `AdminPublicationsFilteringService`](https://docs.google.com/presentation/d/1bp9_EpwnvacRLjioekIEFWgjDCXhjj9echmFN6bClzA/edit?usp=sharing) - in context of `index` action, but this new endpoint will also be used with this filtering service.

## How urgent is a code review?

Within a day or two, as FE will need this endpoint in Iteration 1 of this Epic.

jinjagit committed:
Basic endpoint, auth and related query

Some manual testing - looking good.
To do: add specs.

jinjagit committed:
Test of endpoint - not yet working correctly

jinjagit committed:
fix test

jinjagit committed:
Use if Citizenlab.ee? else in test

jinjagit committed:
Use correct counts of draft status in test

jinjagit committed:
Remove unused parameter declarations

jinjagit committed:
Improve comments

adessy commented on back/app/policies/admin_publication_policy.rb:
```diff
@@ -26,6 +26,10 @@ def reorder?
   end
+
+  def status_counts
+    user&.active?
```
Lill' shortcut :-) 
```suggestion
    active?
```

adessy commented on back/config/routes.rb:
```diff
@@ -139,8 +139,10 @@
+      
       resources :admin_publications, only: %i[index show] do
         patch 'reorder', on: :member
+        get 'status_counts', on: :collection
```
(tiny: I would either go for `counts` only or for `filter_counts` for the sake of consistency -- even `filter` doesn't bring much IMO)

jinjagit replied:
Not sure I understand. The naming is meant to help the reader understand the purpose: to count the `publication_status` of publications. Maybe `publication_status_counts` would be even better?

adessy replied:
Ah yeah, my bad! I was completely off 😑

jinjagit replied:
All good. Really useful feedback in your review. Thanks!

adessy commented on back/app/controllers/web_api/v1/admin_publications_controller.rb:
```diff
@@ -33,6 +33,21 @@ def reorder
   end
 
+  # For use by homepage to get status counts of visible top-level (not in folder) admin_publications,
+  # using: .../web_api/v1/admin_publications/status_counts?remove_not_allowed_parents=true&depth=0
```
(I think that's the kind of comment that can quickly become outdated.)

jinjagit replied:
Hmm. Maybe I will just delete this comment. [Done]

adessy commented on back/app/controllers/web_api/v1/admin_publications_controller.rb:
```diff
@@ -33,6 +33,21 @@ def reorder
+    # For non-admins: publication_status: :draft count should always be zero, and thus not included in response
+    counts = publications.group(:publication_status).count
+
+    authorize :admin_publication, :status_counts
```
Moving the authorization to the top would be more efficient if access is denied.

adessy commented on back/app/controllers/web_api/v1/admin_publications_controller.rb:
```diff
@@ -33,6 +33,21 @@ def reorder
+    publications = policy_scope(AdminPublication.includes(:parent))
+    publications = publication_filterer.filter(publications, params)
+
+    # For non-admins: publication_status: :draft count should always be zero, and thus not included in response
```
Took me a while to understand this comment. Am I right saying that depending on the filters and status of publications on the platform, some publication status might not be shown in the response?

jinjagit replied:
`counts = publications.group(:publication_status).count` will not return a key, nor a value, when 0 records found with any specific `publication_status`.

I guess I'm just pointing out that since non-admins never have permission to see `draft` publications, there will never be `draft` key included in the response.

adessy replied:
But it also applies to situations where for instance all publications on the platform have the same status. Non-represented status would not appear in the response, right?

jinjagit replied:
Yup. I think you're right. I should either make the comment say that, or delete it.

adessy commented on back/spec/acceptance/admin_publications_spec.rb:
```diff
@@ -34,6 +34,7 @@
       parameter :folder, "Filter by folder (project folder id)", required: false
```
I surprised this parameter is not conditioned by `CitizenLab.ee?` since folders are a commercial feature.

jinjagit replied:
I'd have to use a block, like this
```
if Citizenlab.ee?
  parameter :folder, "Filter by folder (project folder id)", required: false
end
```
Because ...
```
parameter :folder, "Filter by folder (project folder id)", required: false if Citizenlab.ee?
```
would apply the conditional to the `required` statement, right?

jinjagit replied:
Done.

adessy reviewed (approved):
Looks good to me! I just made a few unimportant remarks here and there.

jinjagit committed:
Merge branch 'master' of github.com:CitizenLabDotCo/citizenlab into CL2-6833-i1-get-admin-publication-status-counts

jinjagit committed:
Simplify syntax in policy

jinjagit committed:
Remove comments

jinjagit committed:
Move authorization to top of method

jinjagit committed:
Only declare :folder parameter if Citizenlab.ee

Nothing was breaking, but this is probably good practice, as
pointed out by @adessy.
